### ISSUE TITLE
Incorrect warning with -Wabsolute-value

### ISSUE THREAD

**Issue by yt87** [llvm/llvm-project#56976]
This issue is likely related to #25699.

OS and clang versions:
```
$ clang --version
clang version 13.0.0 (Fedora 13.0.0-3.fc35)
Target: x86_64-redhat-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```
Test code:
```
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>

int main(int ac, char **av)
{
    unsigned int i = 10;
    unsigned int j = 11;
    bool b;

    b = abs(i - j) < 2;
    printf("%d\n", b);
    b = i - j < 2;
    printf("%d\n", b);
    return 0;
}
```
Compilation:
```
$ clang -Wabsolute-value foo.c
foo.c:11:9: warning: taking the absolute value of unsigned type 'unsigned int' has no effect [-Wabsolute-value]
    b = abs(i - j) < 2;
        ^
foo.c:11:9: note: remove the call to 'abs' since unsigned values cannot be negative
    b = abs(i - j) < 2;
        ^~~
1 warning generated.
```
Result:
```
$ ./a.out
1
0
```
Obviously the suggestion to remove the call to ``abs`` is wrong.

**Comment by shafik** [comment]
Due to the [Usual Arithmetic Conversions](https://en.cppreference.com/w/c/language/conversion#Usual_arithmetic_conversions) the result of `i-j` is indeed `unsigned` and therefore the diagnostic is correct.

We can see this using the following code:

```c
#include <stdio.h>

void funsigned(unsigned) { printf("unsigned\n");}
void fsigned(int) { printf("signed\n");}

#define f(X) _Generic((X), \
               unsigned int: funsigned, \
               int : fsigned \
              )(X)

int main() {
    unsigned int i = 10;
    unsigned int j = 11;

    f((i-j));
}
```

See it live on godbolt: https://godbolt.org/z/za6oMsrc9

Note that converting a small signed value will result in a very large unsigned value. Also note that `abs(...)` returns an `int` you will see the same result from `b = (int)(i - j) < 2;`

**Comment by yt87** [comment]
It is nitpicking, but I can't resist.
The warning is correct. If I had labs() instead of abs(), the following
note would be true as well. However abs() changes semantics, converts
unsigned to int first, hence the suggestion, if followed blindly, changes
behaviour of the code. Of course, the issue is that taking the difference
of 2 unsigned values without explicit casting is suspicious.
